### ISSUE TITLE
Need help getting it working on Windows XP.

### ISSUE THREAD

**Issue by lss4** [maharmstone/btrfs#378]
I'm trying to install the most recent WinBtrfs 1.7.7 on Windows XP. During installation, it asks me about whether or not continue to install "Storage Volumes". I chose "Continue anyway" to proceed with the installation.

However, the driver doesn't seem to run, despite the files have been installed correctly, and the registry for the btrfs service is indeed present.

I also can't find anything about it in the Event Viewer. It's as if Windows never tried loading that driver to begin with. I read the FAQ and tried changing the partition type to `7` from `83` but it doesn't work either. All it did is making these drives visible to the system, but as "RAW" format.

Is there any way I can diagnose the issue? By the way, I also have Windows 7 dual-booting on that system, and everything works as intended on Win7. Also, is changing partition type really mandatory on Windows XP?

**Comment by maharmstone** [comment]
>is changing partition type really mandatory on Windows XP?

Very yes. The driver doesn't even get told about the partition otherwise.

Does `sc query btrfs` say that the driver is running? If not, what happens when you run `sc start btrfs`?

Bear in mind that I don't habitually test on XP, so it's not out of the question that I've inadvertently broken it...

**Comment by lss4** [comment]
> > is changing partition type really mandatory on Windows XP?
> 
> Very yes. The driver doesn't even get told about the partition otherwise.
> 
> Does `sc query btrfs` say that the driver is running? If not, what happens when you run `sc start btrfs`?
> 
> Bear in mind that I don't habitually test on XP, so it's not out of the question that I've inadvertently broken it...

`sc query btrfs` reports it as stopped. `sc start btrfs` will fail complaining "The specified procedure could not be found".

I checked `btrfs.sys` using Dependency Walker and it seems the problem is that XP's `NTOSKRNL.EXE` doesn't export the symbol `_chkstk` (marked in red) which the driver appears to import.

**Comment by maharmstone** [comment]
Oops, looks like I forgot to reapply 00cf4e7ff3ee8c7e0f06bb1ffe1f9789f2a70f06 when I last upgraded zstd. I'll get it fixed for the next version.

**Comment by lss4** [comment]
The issue persists with 1.7.8 version.

The error's the same and Dependency Walker is reporting the same results (missing `_chkstk`).

**Comment by maharmstone** [comment]
Oh for goodness' sake, it's snuck it into the Release version! Try the Debug version instead.

**Comment by lss4** [comment]
The debug version works. However, I don't know how to change drive letters without using the shell extension as it's not working. I tried changing the sequences in REGEDIT but it doesn't work either.

Checking `shellbtrfs.dll` with Dependency Walker showed the absence of the following symbols in `KERNEL32.DLL`:

```
CompareStringEx
GetLocaleInfoEx
LCMapStringEx
```

**Comment by lss4** [comment]
Version 1.7.8.1 (release) works. However, `shellbtrfs.dll` still doesn't work. This version of `shellbtrfs.dll` imports the following symbols that are absent in Windows XP.

KERNEL32.DLL:
```
CompareStringEx
GetLocaleInfoEx
LCMapStringEx
InitializeCriticalSectionEx (this is new compared to 1.7.8 debug version)
```
MPR.dll:
```
WNetRestoreConnectionA
```

**Comment by maharmstone** [comment]
None of these are called directly, it's all stuff that MSVC has snuck in. I suspect it's going to be increasingly difficult to keep XP compatibility with the userspace tools.

**Comment by lss4** [comment]
> None of these are called directly, it's all stuff that MSVC has snuck in. I suspect it's going to be increasingly difficult to keep XP compatibility with the userspace tools.

This is indeed concerning as one of the intended target of this software is ReactOS, which targets XP/2K3 compatibility.

Don't know if there are any workarounds or alternative toolchains that can help with this situation, if MSVC itself is the problem.

**Comment by maharmstone** [comment]
> This is indeed concerning as one of the intended targets of this software is ReactOS

That's not the case - the target of this software is Windows and only Windows. I've made it very clear to the ReactOS people that any problems they have is their issue.

You might want to try compiling shellbtrfs.dll with mingw, AFAIK that will still work with XP.

**Comment by lss4** [comment]
I'm building this from Manjaro Linux (Arch-based), with mingw-w64 stuffs from official repo installed. Trying to cross-compile for Windows x86 is unfortunately not straightforward.

1. `mingw-x86.cmake` is broken as it incorrectly specified the `x86_64-w64` toolchain. Changing every mention of `x86_64-w64` to `i686-w64` allowed cmake to complete.
2. I can't build everything as some code/config changes are needed. For example, `ubtrfs.c` failed with a missing `ata.h` which I don't know where it is, so I couldn't proceed any further.
3. I need to change `<mountmgr.h>` to `<ddk/mountmgr.h>` in order to complete building just `shellbtrfs.dll`, as the header in question resides in `/usr/i686-w64-mingw32/include/ddk` and the include path is relative to `/usr/i686-w64-mingw32/include`.

Sadly even after all those efforts, the `shellbtrfs.dll` I built still doesn't work, because of missing symbol `GetTickCount64`. However, the other missing symbols I mentioned before is not present in the MinGW-built dll. On the other hand, the missing symbol in MPR.dll is not relevant as when I checked ubtrfs.dll with Dependency Walker it also had that one, but did not give me an error.

Googling about the missing symbol in question pointed me to [this MSYS2 issue](https://github.com/msys2/MINGW-packages/issues/5139) which is probably relevant.

So unless I try building it with an old enough MinGW environment (or the other around, someone to actually extend the XP kernel in whatever way to implement the function in question), getting `shellbtrfs.dll` to work on XP is not going to be easy.

**Comment by lss4** [comment]
Okay I finally got shellbtrfs working on XP.

1. Make a Ubuntu 18.04 (Bionic) VM. A simple one should do.
2. Install mingw-w64 and related stuffs in the VM. The MinGW shipped in Bionic is around 5.0.3 which can be used for the purpose.
3. Install _latest_ CMake (the cmake shipped in Bionic is 3.10, but btrfs requires 3.15 or higher). You can use [Kitware's repo](https://apt.kitware.com/) (see [this guide](https://askubuntu.com/a/865294)), or maybe from snap.
4. Make changes according to my previous post (fix `mingw-x86.cmake`, change a few includes of `<mountmgr.h>` to `<ddk/mountmgr.h>`).
5. Create a build folder (e.g. `./build`), `cd` into the build folder and run these.
```
cmake -DCMAKE_TOOLCHAIN_FILE=../mingw-x86.cmake ..
make shellbtrfs
```
6. If nothing goes wrong, take the resulted shellbtrfs.dll and replace the existing one. The DLL built using this old environment will not have the problem of missing `GetTickCount64` symbol and will run on XP.

Just a side note, when using the Btrfs shell extension to change drive letter, you may be prompted to choose which user to run. You can run as current user if it's already administrator, **but be sure to uncheck "Protect my computer and data from unauthorized program activity"** or you'll get an Access Denied error.

On the other hand, it seems on WinXP the shell extension can't properly show disk sizes in Volume Usage and other places. I got some strange `H*7` and `?7`. Maybe it's related to some linker warnings I got during the build process, but it's not important at the moment as I mainly need the shell extension working to be able to change the drive letter and it can already achieve that.

```
Warning: resolving _AddDeviceW by linking to _AddDeviceW@16
Use --enable-stdcall-fixup to disable these warnings
Use --disable-stdcall-fixup to disable these fixups
Warning: resolving _CreateSnapshotW by linking to _CreateSnapshotW@16
Warning: resolving _CreateSubvolW by linking to _CreateSubvolW@16
Warning: resolving _DllCanUnloadNow by linking to _DllCanUnloadNow@0
Warning: resolving _DllGetClassObject by linking to _DllGetClassObject@12
Warning: resolving _DllInstall by linking to _DllInstall@8
Warning: resolving _DllRegisterServer by linking to _DllRegisterServer@0
Warning: resolving _DllUnregisterServer by linking to _DllUnregisterServer@0
Warning: resolving _PauseBalanceW by linking to _PauseBalanceW@16
Warning: resolving _RecvSubvolGUIW by linking to _RecvSubvolGUIW@16
Warning: resolving _RecvSubvolW by linking to _RecvSubvolW@16
Warning: resolving _ReflinkCopyW by linking to _ReflinkCopyW@16
Warning: resolving _RemoveDeviceW by linking to _RemoveDeviceW@16
Warning: resolving _ResetStatsW by linking to _ResetStatsW@16
Warning: resolving _ResizeDeviceW by linking to _ResizeDeviceW@16
Warning: resolving _SendSubvolGUIW by linking to _SendSubvolGUIW@16
Warning: resolving _SendSubvolW by linking to _SendSubvolW@16
Warning: resolving _ShowChangeDriveLetterW by linking to _ShowChangeDriveLetterW@16
Warning: resolving _ShowPropSheetW by linking to _ShowPropSheetW@16
Warning: resolving _ShowScrubW by linking to _ShowScrubW@16
Warning: resolving _StartBalanceW by linking to _StartBalanceW@16
Warning: resolving _StartScrubW by linking to _StartScrubW@16
Warning: resolving _StopBalanceW by linking to _StopBalanceW@16
Warning: resolving _StopScrubW by linking to _StopScrubW@16
```

**Comment by lss4** [comment]
I'm closing the issue as I've got my answers of all related questions.
